include/coap/pdu.h

```diff
@@ -308,6 +417,15 @@ void coap_pdu_clear(coap_pdu_t *pdu, size_t size);
  */
 coap_pdu_t *coap_new_pdu(void);
 
+/**
+ * Creates a new CoAP PDU. The object is created on the heap and must be released
+ * using coap_delete_pdu();
+ *
+ * @deprecated This function allocates the maximum storage for each
+ * PDU. Use coap_pdu_init2() instead.
+ */
+coap_pdu_t *coap_new_pdu2(coap_transport_t transport, unsigned int size);
```

Comment:
Regarding the general transport-specific API: I think that it is not a good idea to specify the transport for each and every operation on a PDU. When only UDP existed as transport for CoAP, you could safely assume that the CoAP message under construction will have a certain serialization on the wire. Now, when introducing multiple transports, the serialization should happen immediately when writing the message onto the network. The API for setting the token, adding options or data, etc., should not have to care which serialization will be used,